baseband/guppi/header.py

```diff
@@ -0,0 +1,375 @@
+            self.comments[key.upper()] = comment
+
+        super(GUPPIHeader, self).__setitem__(key.upper(), value)
+
+    @property
+    def nbytes(self):
+        """Size of the header in bytes."""
+        return (len(self) + 1) * 80
+
+    @property
+    def payload_nbytes(self):
+        """Size of the payload in bytes."""
+        return self['BLOCSIZE']
+
+    @payload_nbytes.setter
+    def payload_nbytes(self, payloadsize):
+        self['BLOCSIZE'] = payloadsize
+
+    @property
+    def frame_nbytes(self):
+        """Size of the frame in bytes."""
+        return self.nbytes + self.payload_nbytes
+
+    @frame_nbytes.setter
+    def frame_nbytes(self, frame_nbytes):
+        self.payload_nbytes = frame_nbytes - self.nbytes
+
+    @property
+    def bps(self):
+        """Bits per elementary sample."""
+        return self['NBITS']
+
+    @bps.setter
+    def bps(self, bps):
+        self['NBITS'] = bps
+
+    @property
+    def complex_data(self):
+        """Whether the data are complex."""
+        return self['OBSNCHAN'] != 1
+
+    @property
+    def npol(self):
+        """Number of polarisations."""
+        return self['NPOL'] // (2 if self.complex_data else 1)
+
+    @npol.setter
+    def npol(self, npol):
+        self['NPOL'] = npol * (2 if self.complex_data else 1)
+
+    @property
+    def nchan(self):
+        """Number of channels."""
+        return self['OBSNCHAN']
+
+    @nchan.setter
+    def nchan(self, nchan):
+        self['OBSNCHAN'] = nchan.__index__()
+
+    @property
+    def sample_shape(self):
+        """Shape of a sample in the payload (npol, nchan)."""
+        return self.npol, self.nchan
+
+    @sample_shape.setter
+    def sample_shape(self, sample_shape):
+        # Need to set nchan first to properly set npol, since nchan is
+        # connected to complex_data.
+        self.nchan = sample_shape[1]
+        self.npol = sample_shape[0]
+
+    @property
+    def _bpcs(self):
+        """Bits per complete sample."""
+        # OBSNCHAN includes factor of 2 for real/complex components.
+        return self['OBSNCHAN'] * self['NPOL'] * self.bps
+
+    @property
+    def sample_rate(self):
+        """Number of complete samples per second.
+
+        Can be set with a negative quantity to set `sideband`.
+        """
+        return (1. / self['TBIN']) * u.Hz
+
+    @sample_rate.setter
+    def sample_rate(self, sample_rate):
+        self['TBIN'] = 1. / abs(sample_rate.to_value(u.Hz))
+        bw = (sample_rate.to_value(u.MHz) * self['OBSNCHAN'] /
+              (1 if self.complex_data else 2))
+        self['OBSBW'] = (-1 if self.get('OBSBW', bw) < 0 else 1) * bw
+
+    @property
+    def sideband(self):
+        """True if upper sideband."""
+        return self['OBSBW'] > 0
+
+    @sideband.setter
+    def sideband(self, sideband):
+        self['OBSBW'] = (1 if sideband else -1) * abs(self['OBSBW'])
+
+    @property
+    def time_ordered(self):
+        return self['PKTFMT'] != 'SIMPLE'
```

Comment:
Yeah, this kind of bothered me too.  I'll change it since with `verify` we only allow `'SIMPLE'` or `'1SFA'`, and added a warning in the comments that supporting other methods will require some header revisions.  On the plus side this also makes one of the tests a bit more straightforward.

Though, why is `header.time_ordered` True when it is `1SFA`?  I know DSPSR uses that logic, but `1SFA` implies channel-first ordering...